### PR TITLE
Initial support for model variants (bert, albert/masked_lm)

### PR DESCRIPTION
**Pull request by kmabeeTT**
FYI - @AleksKnezevic @mrakitaTT @nvukobratTT 

Initial model variant support for Bert and Albert Masked LM

 - This will be example for porting more models that have variants
 - Some infra for defining multiple optional variants for models and
   allowing user/test to query available variants, and request variants
 - port albert_masked_lm with base, large, xlarge, xxlarge variants from tt-torch
 - tested out changes on tt-torch, queried variants can be used as pytest params
 
==
 
This is to add support for models that have multiple variants. Want to get some feedback on this to settle on something decent so we can go ahead and port all our models to tt-forge-models starting this week. The way I have it setup is there is some shared variant logic in the base class, such that model variants are optional.

In tt-torch, we use variants like this (example from `tests/models/albert/test_albert_masked_lm.py`):

```

# Print available variants for reference
available_variants = ModelLoader.query_available_variants()
print("Available variants:", available_variants)

<snip>

# Automatically discover variants and use to populate pytest params of available variants:
@pytest.mark.parametrize(
    "variant_info",
    available_variants.items(),
    ids=list(available_variants.keys()),
)

<snip>

# Use variant when loading model and inputs:

class ThisTester(ModelTester):
    def _load_model(self):
        return ModelLoader.load_model(
            variant=self.variant, dtype_override=torch.bfloat16
        )

    def _load_inputs(self):
        self.inputs = ModelLoader.load_inputs(
            variant=self.variant, dtype_override=torch.bfloat16
        )

Can query as follows currently:

print("Available ALBERT variants:", ModelLoader.query_available_variants())

```

Corresponding tt-torch changes that use this are on branch `kmabee/model_variants` in PR: https://github.com/tenstorrent/tt-torch/pull/929

### CODE REVIEW

**Comment by kmabeeTT** [comment]
example output snippet from albert test in tt-torch for reference:

```
% pytest -svv tests/models/albert/test_albert_masked_lm.py::test_albert_masked_lm[single_device-full-albert-xxlarge-v2-eval]

<snip>

Available variants: {'albert-base-v2': 'ALBERT base model (12M parameters)', 'albert-large-v2': 'ALBERT large model (18M parameters)', 'albert-xlarge-v2': 'ALBERT xlarge model (60M parameters)', 'albert-xxlarge-v2': 'ALBERT xxlarge model (235M parameters)'}
collected 1 item

<snip>

tests/models/albert/test_albert_masked_lm.py::test_albert_masked_lm[single_device-full-albert-xxlarge-v2-eval] Testing model_name: albert/albert-xxlarge-v2-masked_lm variant_desc: ALBERT xxlarge model (235M parameters)

<snip>

Results for output 0:
  PCC: 0.9878, threshold: 0.98 __
__ (assert_atol == False)

Model: albert/albert-xxlarge-v2-masked_lm | Input: The capital of [MASK] is Paris. | Mask: france
PASSED
```

**Comment by mrakitaTT** [comment]
Hey @kmabeeTT just leaving a comment to let you know that I plan to take a look today, I'm in a bit of a jam right now. We have a similar structure in tt-xla with base tester for the model architecture and child testers for each variant, so I wanted to compare this design with what we have.

File: bert/pytorch/loader.py [hunk 1]
```diff
@@ -13,30 +13,58 @@
 class ModelLoader(ForgeModel):
     """BERT model loader implementation for question answering tasks."""
 
+    # Dictionary of available model variants
+    _VARIANTS = {
+        "base": {
+            "pretrained_model_name": "phiyodr/bert-base-finetuned-squad2",
+            "max_length": 256,
```
[the remark this answers is not in the document]
Comment (mrakitaTT) [reply bert/pytorch/loader.py:20]: >Should this be part of the variant? I would expect variants only to contain variant name.

In order to be able to iterate through variants and to have loader functions work in a generic way, we need to have the variant specific configs stored somewhere, though I agree that we should figure out a better format instead of naked dicts. We also need to store somewhere basic model info for each model that we will use to build model name and write into tags for dashboard.

Maybe we can have these classes:
- `ModelInfo` containing basic model info that we need for dashboard, you can see example [here](https://github.com/tenstorrent/tt-xla/blob/241ac82bcb6ca9433d56204c2350679a2c839442/tests/utils.py#L89) that @kmitrovicTT made in tt-xla in [this PR](https://github.com/tenstorrent/tt-xla/pull/641/files) (not yet checked in because we want to move this to tt-forge-models repo)
- `ModelConfig` containing specific model config needed for variants, like `description`, `pretrained_model_name`, etc. We could also have child classes like `LLMModelConfig` which would contain even more specific things like `max_length`.

Then `_VARIANTS` would become:
```
_VARIANTS: Dict[str, ModelConfig] = {
    "base": LLMModelConfig(
        description="BERT-base fine-tuned on SQuAD v2",
        pretrained_model_name="phiyodr/bert-base-finetuned-squad2",
        max_length=256,
    },
}
```

and we can have base loader method like `get_model_info()` which would look like this (hardcoded strings can be moved to constants in the loader class):
```
@classmethod
def get_model_info(cls, variant=None) -> ModelInfo:
    variant_name = cls._validate_variant(variant)

    return ModelInfo(
        model="albert_v2",
        variant=variant_name,
        group=ModelGroup.GENERALITY,
        task=ModelTask.NLP_MASKED_LM,
        source=ModelSource.HUGGING_FACE,
        framework=Framework.JAX,
    )
```

What do you guys think?

File: albert/masked_lm/pytorch/loader.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        if variant is not None or not hasattr(cls, "tokenizer"):
+            cls.load_model(variant=variant, dtype_override=dtype_override)
+
+        # Create tokenized inputs for the masked language modeling task
+        inputs = cls.tokenizer(cls.sample_text, return_tensors="pt")
+
+        return inputs
+
+    @classmethod
+    def decode_output(cls, outputs, inputs=None):
```
[the remark this answers is not in the document]
Comment (mrakitaTT) [reply albert/masked_lm/pytorch/loader.py:104]: Yeah, once we start adding models there will be a lot of stuff repeating. We will probably need to have `pytorch_loader.py` and `jax_loader.py`, with default implementations for some of internal functions that most of the models can use, and then specific model loaders can override them.

File: albert/masked_lm/pytorch/loader.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    # Default variant to use
+    DEFAULT_VARIANT = "albert-base-v2"
+
+    # Shared configuration parameters
+    sample_text = "The capital of [MASK] is Paris."
+
+    # Track the current variant being used
+    _current_variant = None
```
[the remark this answers is not in the document]
Comment (mrakitaTT) [reply albert/masked_lm/pytorch/loader.py:43]: With the current design with all methods being `classmethod` I would remove it, keeping state between class static methods could just lead to bugs. Though we should maybe consider not making these methods `classmethod`, I cannot think of specific use case right now but if we ever want to have multiple loader instances in the future or to cache some internal state in the loader - we will have many files to change. That way we will also be able to do:
```
loader = ModelLoader(variant="base")
model = loader.load_model()
inputs = loader.load_inputs()
```
instead of repeating the variant name for each method:
```
model = ModelLoader.load_model(variant="base")
inputs = ModelLoader.load_inputs(variant="base")
```

File: bert/pytorch/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvukobratTT) [reply bert/pytorch/loader.py:20]: @mrakitaTT, do you happen to have more examples about variant-specific configs?

[the remark this answers is not in the document]
Comment (mrakitaTT) [reply bert/pytorch/loader.py:20]: Here are some variant-specific differences that I remember:
- Some variants are in RED models group and some in GENERALITY - This should be handled by `get_model_info` though
- We may want to test different model config with different variants, for example BigBird comes with 2 attention implementations: `original_full` & `block_sparse`. For the sequence length < 1024, using `original_full` is advised as there is no benefit in using `block_sparse` attention. We test `sequence length < 1024` + `original_full` attention with BigBird base model, and `sequence length > 1024` + `block_sparse` with BigBird large model.
- Some variants are using different implementation classes, for example BlenderBot small is loaded with `FlaxBlenderbotSmallForConditionalGeneration` while other BlenderBot variants are loaded with `FlaxBlenderbotForConditionalGeneration`.
- From v0.2 version of Mistral-7B sliding window attention was removed, but Transformers Flax implementation of Mistral-7B wasn't updated to take that into account, so we have to manually disable it for that variant.

Some of these could be handled as if statements based on the variant name in the loader, but some are more elegant to keep in some class like `ModelConfig`, like model path and description. I think it wouldn't hurt to setup a class like that anyways which we can use for customizations if necessary and also to avoid naked strings dicts.

[the remark this answers is not in the document]
Comment (nvukobratTT) [reply bert/pytorch/loader.py:20]: @kmabeeTT, what are your thoughts?

File: albert/masked_lm/pytorch/loader.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            "pretrained_model_name": "albert/albert-xxlarge-v2",
+            "description": "ALBERT xxlarge model (235M parameters)",
+        },
+    }
+
+    # Default variant to use
+    DEFAULT_VARIANT = "albert-base-v2"
+
+    # Shared configuration parameters
+    sample_text = "The capital of [MASK] is Paris."
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply albert/masked_lm/pytorch/loader.py:40]: Not sure, should it? It's currently consistent with other models so far like bert, falcon and README.md I added that have input text in `ModelLoader` class variable instead of load_inputs variable.

```
class ModelLoader(ForgeModel):
    """Falcon model loader implementation for question answering tasks."""

    # Shared configuration parameters
    model_name = "tiiuae/Falcon3-1B-Base"
    input_text = "Write a function to calculate the factorial of a number"
```

File: bert/pytorch/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply bert/pytorch/loader.py:20]: Thanks a lot for discussion here - I got to reading this late, and it's way more involved than I was thinking originally, but it makes sense - I don't have any better suggestions right now.  Would like to find minimum possible infra/classes here I can include for the 2 models in my PR, or if I should include a few others too with more interesting `ModelInfo/ModelConfig` I can do that too - I'll try to see what I can figure out here tomorrow in terms of implementation. If you have more hints/tips/suggestions that will help me, feel free to drop here.

[the remark this answers is not in the document]
Comment (kmabeeTT) [reply bert/pytorch/loader.py]: I pushed a commit 55f2462 with some initial changes in spirit of this discussion incorporating ModelInfo, ModelConfig, constant enums, take a look - talk tomorrow.

File: config.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+class ModelGroup(StrEnum):
+    """Model groups for categorization and reporting"""
+
+    GENERALITY = "generality"
+    RED = "red"
+    PRIORITY = "priority"
+
+
+class ModelTask(StrEnum):
```
Comment (mrakitaTT) [inline config.py:27]: Here is a complete list of HF tasks: https://github.com/tenstorrent/tt-xla/blob/main/tests/utils.py#L33

File: albert/masked_lm/pytorch/loader.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+    # Tokenizer shared across instances
+    tokenizer = None
```
Comment (mrakitaTT) [inline albert/masked_lm/pytorch/loader.py:55]: Should be `self.tokenizer` and set to None inside the `__init__` function, tokenizer depends on the variant. This doesn't strictly have to be done since we can just assign `self.tokenizer` in the methods `load_model` and `load_inputs`, but I like to explicitly set the member variables in `__init__` method so that it is obvious which member variables class objects have.

File: albert/masked_lm/pytorch/loader.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        model = AlbertForMaskedLM.from_pretrained(pretrained_model_name, **model_kwargs)
+
+        # Initialize the tokenizer if not already done
+        # Note: tokenizer is shared across all instances
+        tokenizer_kwargs = {}
+        if dtype_override is not None:
+            tokenizer_kwargs["torch_dtype"] = dtype_override
+
+        if ModelLoader.tokenizer is None:
+            ModelLoader.tokenizer = AutoTokenizer.from_pretrained(
```
Comment (mrakitaTT) [inline albert/masked_lm/pytorch/loader.py:105]: `self.tokenizer`, in `load_inputs()` too

File: base.py [hunk 8]
```diff
@@ -1,30 +1,132 @@
...
+            ModelConfig or None: Variant configuration object, or None for models without variants.
+        """
+        variant = cls._validate_variant(variant)
+        if variant is None:
+            return None
+
+        return cls._VARIANTS[variant]
+
+    @classmethod
+    def get_model_info(cls, variant=None) -> ModelInfo:
```
Comment (mrakitaTT) [inline base.py:100]: Could also be changed to not be `classmethod` and use `self._variant` which was already validated

File: albert/masked_lm/pytorch/loader.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+        Args:
+            dtype_override: Optional torch.dtype to override the model inputs' default dtype.
+
+        Returns:
+            dict: Input tensors that can be fed to the model.
+        """
+        # Ensure tokenizer is initialized
+        if ModelLoader.tokenizer is None:
+            self.load_model(dtype_override=dtype_override)
```
Comment (mrakitaTT) [inline albert/masked_lm/pytorch/loader.py:122]: Since you've already made this change I would also extract tokenizer loading into separate function so we don't have to load the whole model here, you can just call that function here and from `load_model` where tokenizer is loaded.

File: config.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+class Framework(StrEnum):
+    """Framework the model is implemented in"""
+
+    JAX = "jax"
+    TORCH = "pytorch"
+    NUMPY = "numpy"
+
+
+@dataclass(frozen=True)
+class ModelInfo:
```
Comment (mrakitaTT) [inline config.py:76]: Let's add parallelism too

**Review by mrakitaTT — APPROVED** [review]
Looks awesome now, thanks Kyle!

File: base.py (hunk 8 above)
Comment (mrakitaTT) [reply base.py:100]: Actually this will make it difficult to log test properties without loader instance, let's leave it as a class method for now

File: config.py (hunk 10 above)
Comment (kmabeeTT) [reply config.py:82]: Done added this - FYI @jameszianxuTT / @nvukobratTT 

```
class Parallelism(StrEnum):
    """Multi-device parallelism strategy the model is using."""

    SINGLE_DEVICE = "single_device"
    DATA_PARALLEL = "data_parallel"
    TENSOR_PARALLEL = "tensor_parallel"
```

File: albert/masked_lm/pytorch/loader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrakitaTT) [reply albert/masked_lm/pytorch/loader.py:48]: Since it is used only inside `load_inputs` I'd move it to be defined there inside the function

File: albert/masked_lm/pytorch/loader.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    def get_model_info(cls, variant=None) -> ModelInfo:
+        """Get model information for dashboard and metrics reporting.
+
+        Args:
+            variant: Optional string specifying which variant to get info for.
+
+        Returns:
+            ModelInfo: Information about the model and variant
+        """
+        variant = cls._validate_variant(variant)
```
Comment (mrakitaTT) [inline albert/masked_lm/pytorch/loader.py:63]: In order to avoid this line in every loader, maybe we can do something like:
```
# base.py

@classmethod
def get_model_info(cls, variant=None) -> ModelInfo:
    variant_name = cls._validate_variant(variant)
    return cls._get_model_info(variant_name)

@classmethod
def _get_model_info(cls, variant_name: str) -> ModelInfo:
    raise NotImplementedError("Subclasses must implement _get_model_info")


# loader.py

@classmethod
def _get_model_info(cls, variant_name: str) -> ModelInfo:
    return ModelInfo(
        ...
    )
```

File: albert/masked_lm/pytorch/loader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply albert/masked_lm/pytorch/loader.py]: Thx, opened issue for this here: https://github.com/tenstorrent/tt-forge-models/issues/25

File: albert/masked_lm/pytorch/loader.py (hunk 7 above)
Comment (kmabeeTT) [reply albert/masked_lm/pytorch/loader.py]: Done, resolving.

File: albert/masked_lm/pytorch/loader.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    LLMModelConfig,
+    ModelInfo,
+    ModelGroup,
+    ModelTask,
+    ModelSource,
+    Framework,
+)
+
+
+class ModelLoader(ForgeModel):
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply albert/masked_lm/pytorch/loader.py:21]: I accidentally deleted Nikola's comment which used to be above (sorry! Was tricked, thinking it was duplicately posted, but it wasn't). Adding it back:

> What is the value of introducing enums for every model variant?
> 
> In the majority of cases, this will be a 1-1 mapping. I.e., we'll have # enums == # variants. This looks like additional boilerplate code that we need to maintain.

**Comment by kmabeeTT** [comment]
Thanks guys for feedback! I don't have any CNN or ONNX models handy that I could comfortably/easily port without follow up questions, so rather than delay this another 1-2 days I better merge now. Opened a couple follow up issues for things discussed above that were deferred.
